### ISSUE TITLE
sporadic git command failed "git write-tree"

### ISSUE THREAD

**Issue by petergaultney** [jacob-meacham/serverless-plugin-git-variables#25]
For a while I'd thought this was because I was doing simultaneous deployments, but recently it has happened when only running a single deployment.

```
$ sls deploy

  Error --------------------------------------------------

  Command failed: git write-tree
fatal: Unable to create '/Users/peter/workspace/my-serverless-app/.git/index.lock': File exists.
```

A second run without any intervention on my part (I did not look for or delete an index.lock file) succeeded, which suggests to me that there is actually some kind of race condition inside this plugin?

**Comment by jacob-meacham** [comment]
@petergaultney interesting, this isn't something I've seen myself - I'll take a look

**Comment by iamakulov** [comment]
Can confirm that this happens to me too – although on every run, not just on the first one.

Unlike @petergaultney, I’m using Windows.

**Comment by iamakulov** [comment]
Actually, I think this is related to this snippet:

https://github.com/jacob-meacham/serverless-plugin-git-variables/blob/a6a33f5acd1e06a34835c4c623ec0bb745a5cda7/src/index.js#L99-L105

I tried rewriting the snippet locally to use sequential `await` instead of `map` & `Promise.all`, and the error stopped happening.

Let me submit a PR.

**Comment by atz** [comment]
Seeing this intermittently (but frequently) on MacOSX Mojave, too.

**Comment by amanagrawal9** [comment]
When it will be fixed???

**Comment by jacob-meacham** [comment]
Release as 3.3.2

**Comment by james-hu** [comment]
Maybe it is not actually resolved? I already upgraded to 3.3.2 and still see this:
```
[2019-07-14T23:37:47.166Z]   Error --------------------------------------------------

[2019-07-14T23:37:47.166Z]  

[2019-07-14T23:37:47.166Z]   Command failed: git write-tree

[2019-07-14T23:37:47.166Z] fatal: Unable to create '/home/jenkins/workspace/s_lead-referral-services_develop/.git/index.lock': File exists.

[2019-07-14T23:37:47.166Z] 

[2019-07-14T23:37:47.166Z] Another git process seems to be running in this repository, e.g.

[2019-07-14T23:37:47.166Z] an editor opened by 'git commit'. Please make sure all processes

[2019-07-14T23:37:47.166Z] are terminated then try again. If it still fails, a git process

[2019-07-14T23:37:47.166Z] may have crashed in this repository earlier:

[2019-07-14T23:37:47.166Z] remove the file manually to continue.

```

BTW, I run several ```sls deploy``` concurrently from the same git repo in different directories for different Lambda functions.

A backoff-then-retry may resolve the issue, I think.

**Comment by petergaultney** [comment]
> BTW, I run several `sls deploy` concurrently from the same git repo in different directories for different Lambda functions.

Yeah, if you're running these in parallel outside the plugin it would still be possible to have the failure.

> A backoff-then-retry may resolve the issue, I think.

It would be very nice if the plugin would handle this. More to the point, it seems weird to me that this is not a read-only operation.

What if instead of doing `git write-tree` and then `git diff-index`, the plugin did either `_exec('git status --porcelain 2>/dev/null| egrep "^(M| M)" | wc -l') > 0`  (for both uncommitted and untracked) or simply `_exec('git diff --stat') !== ''` (for uncommitted but ignoring untracked files). Then this would neatly skip the entire issue even in parallel deployment situations?

**Comment by jacob-meacham** [comment]
@petergaultney I like this suggestion - I'd like to avoid your first idea because that makes a number of assumptions about the environment (and would fail on Windows in particular), but the second feels very reasonable. I'll have to think a bit about whether I want to release this as a breaking change or not but I'm earmarking a bit of time next week to plow through some of the issues that have piled up in this project

**Comment by petergaultney** [comment]
I've tried out the PR referenced above from @alanning and it doesn't seem to solve the core problem, which is that the underlying implementation to get the git dirty status is not read-only with respect to `git` itself. 

I'm considering trying to fix this in our own fork. If/when I do, I'll make a PR.

**Comment by navrkald** [comment]
Hi @jacob-meacham we are still facing this issue. For more info, pls see PR: https://github.com/jacob-meacham/serverless-plugin-git-variables/pull/55

**Comment by james-hu** [comment]
@jacob-meacham could we re-open this? The issue is still unsolved and there are two PRs related to this one.

**Comment by jacob-meacham** [comment]
@james-hu sure thing - I've scheduled some time this week to take a look. Thanks!

**Comment by james-hu** [comment]
@jacob-meacham I've created a new pull request for this https://github.com/jacob-meacham/serverless-plugin-git-variables/pull/62  but I don't know how to trigger the test and coverage check. Could you have a look and see if this can be merged?

**Comment by jacob-meacham** [comment]
Released a fix in 4.1.0. Thanks!
